### PR TITLE
Utilize new memory features (ReadOnlySpan, BinaryPrimitives, and Span.CopyTo)

### PR DESCRIPTION
**Pull request by iamcarbon**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description

Updates to utilize new memory features:

1. Remove ReverseBytes helpers & replace with BinaryPrimitives
2. Removes the BitConverter abstractions from EndianBinaryReader&Writer 
3. Use ReadOnlySpan when passing source data to methods
4. Replace custom logic in SpanHelper.Copy with Span.CopyTo. @antonfirsov | Were there any specific optimizations in the custom SpanHelper.Copy logic that we will miss? It looks like a lot of work went into improving the performance of Span.CopyTo(...) within the 2.1 cycle. There's only a few call sites left to update before removing this all together.
5. Use BinaryPrimitives for reading color profiles
6. Removes the custom BitConverters 
7. Eliminate allocating an array and encoding the MagicBytes & ApplicationIdentifiers in GIFEncode for each image. This eliminates the last use of the text encoding concerns within the EndianBinaryReader&Writer.
8. Remove text encoding concerns from EndianBinaryReader&Writer
9. Create specialized BigEndianBinaryWriter & LittleEndianBinaryWriter classes

### CODE REVIEW

**Comment by iamcarbon** [comment]
Ready for review! I tried to focus just on replacing the Endian logic, but did end up updating some other things too.

We should also consider exposing the LoadPixelData(Configuration, Span<TPixel>, ...) method and removing the Array overload if we do. We're copying the data to memory under our control so we don't need to worry about the lifetime of the original bytes.

I'll focus these PR's in the future. Hoping to update all the call sites that would benefit from in next!

File: src/ImageSharp/Formats/Png/PngEncoderCore.cs [hunk 1]
```diff
@@ -243,30 +245,17 @@ public void Dispose()
...
-
         /// <summary>
         /// Writes an integer to the stream.
         /// </summary>
         /// <param name="stream">The <see cref="Stream"/> containing image data.</param>
         /// <param name="value">The value to write.</param>
         private static void WriteInteger(Stream stream, int value)
         {
-            byte[] buffer = BitConverter.GetBytes(value);
+            byte[] buffer = new byte[4];
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/PngEncoderCore.cs]: `stackalloc` + `Span<byte>` ?

File: src/ImageSharp/Formats/Gif/GifConstants.cs [hunk 2]
```diff
@@ -21,6 +21,11 @@ internal static class GifConstants
         /// </summary>
         public const string FileVersion = "89a";
 
+        /// <summary>
+        /// The ASCII encoded bytes used to identify the GIF file.
+        /// </summary>
+        internal static readonly byte[] MagicNumber = { 71, 73, 70, 56, 57, 97 }; // GIF89a
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Gif/GifConstants.cs]: 👍 A good, simple win. We could use `Encoding.UTF8.GetBytes(FileType + FileVersion)` though to make it a little less "Magic"

File: src/ImageSharp/Formats/Png/PngDecoderCore.cs [hunk 3]
```diff
@@ -1159,22 +1157,19 @@ private void ReadTextChunk(ImageMetaData metadata, byte[] data, int length)
         /// <summary>
         /// Reads a header chunk from the data.
         /// </summary>
-        /// <param name="data">The <see cref="T:byte[]"/> containing  data.</param>
-        private void ReadHeaderChunk(byte[] data)
+        /// <param name="data">The <see cref="T:Span{byte}"/> containing  data.</param>
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/PngDecoderCore.cs]: `The <see cref="ReadOnlySpan{byte}"/>`?

File: src/ImageSharp/Image.LoadPixelData.cs [hunk 4]
```diff
@@ -103,13 +98,9 @@ private static Image<TPixel> LoadPixelData<TPixel>(Configuration config, Span<by
...
-            int count = width * height;
-            Guard.MustBeGreaterThanOrEqualTo(data.Length, count, nameof(data));
-
-            var image = new Image<TPixel>(config, width, height);
-            SpanHelper.Copy(data, image.GetPixelSpan(), count);
-
-            return image;
+            // There's an implict cast to Span from Array
+            // Should we remove this overload and expose Span<TPixel> ?
+            return LoadPixelData(config, new Span<TPixel>(data), width, height);
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Image.LoadPixelData.cs]: See #489 also

**Comment by JimBobSquarePants** [comment]
@iamcarbon I've just added benchmarks to the solution. Here's pre-PR ones for decoding Png and Gif

Method |     TestImage |       Mean |     Error |    StdDev | Scaled | ScaledSD |   Gen 0 | Allocated |
--------------------- |-------------- |-----------:|----------:|----------:|-------:|---------:|--------:|----------:|
'System.Drawing Gif' | Gif/rings.gif |   438.9 us |  51.44 us |  2.906 us |   1.00 |     0.00 | 51.2695 | 105.44 KB |
'ImageSharp Gif' | Gif/rings.gif | 1,284.6 us | 185.30 us | 10.470 us |   2.93 |     0.03 |       - |    2.1 KB |



Method |      TestImage |     Mean |     Error |    StdDev | Scaled | ScaledSD |   Gen 0 |   Gen 1 |   Gen 2 | Allocated |
--------------------- |--------------- |---------:|----------:|----------:|-------:|---------:|--------:|--------:|--------:|----------:|
'System.Drawing Png' | Png/splash.png | 3.056 ms | 0.3429 ms | 0.0194 ms |   1.00 |     0.00 | 74.2188 | 74.2188 | 74.2188 | 240.54 KB |
'ImageSharp Png' | Png/splash.png | 7.610 ms | 1.2093 ms | 0.0683 ms |   2.49 |     0.02 | 46.8750 |       - |       - |  97.21 KB |

**Comment by iamcarbon** [comment]
@JimBobSquarePants Thanks for fixing the benchmarks. I'll give them a try tomorrow and see if there are any improvements or regressions switching to the native Span.CopyTo methods.

**Comment by JimBobSquarePants** [comment]
@iamcarbon Don't worry about the Colorspace stuff just now in this PR, it'll get too big. 

I have a cunning plan to make them `IPixel` compatible anyway 😮

**Comment by iamcarbon** [comment]
Agreed. There's a big chunk of stuff here already. I'll focus on benchmarking next and fix up any regressions.

**Comment by JimBobSquarePants** [comment]
Good man!

**Comment by iamcarbon** [comment]
The difference between using the SpanHelper.Copy & Span.CopyTo in Slow Span (.NET 4.7) looks to be within the margin of error and we didn't regress on any of the other changes on these two benchmarks.

We should start seeing significant improvements when running on .NETCORE 2.1 and when fast span lands in the desktop framework.

### DecodeGif

Baseline, PRE PR (Upstream / Master)  —

```
               Method |     TestImage |       Mean |      Error |    StdDev | Scaled | ScaledSD |   Gen 0 | Allocated |
--------------------- |-------------- |-----------:|-----------:|----------:|-------:|---------:|--------:|----------:|
 'System.Drawing Gif' | Gif/rings.gif |   423.3 us |   263.6 us |  14.89 us |   1.00 |     0.00 | 16.6016 | 105.37 KB |
     'ImageSharp Gif' | Gif/rings.gif | 1,246.0 us | 1,902.2 us | 107.48 us |   2.95 |     0.22 |       - |   2.11 KB |
```

**PR** —

using SpanHelper.Copy

```
               Method |     TestImage |       Mean |      Error |   StdDev | Scaled | ScaledSD |   Gen 0 | Allocated |
--------------------- |-------------- |-----------:|-----------:|---------:|-------:|---------:|--------:|----------:|
 'System.Drawing Gif' | Gif/rings.gif |   412.5 us |   305.7 us | 17.27 us |   1.00 |     0.00 | 16.6016 | 105.37 KB |
     'ImageSharp Gif' | Gif/rings.gif | 1,372.0 us | 1,079.4 us | 60.99 us |   3.33 |     0.16 |       - |   2.11 KB |
```

using Span.CopyTo

```
               Method |     TestImage |       Mean |      Error |     StdDev | Scaled | ScaledSD |   Gen 0 | Allocated |
--------------------- |-------------- |-----------:|-----------:|-----------:|-------:|---------:|--------:|----------:|
 'System.Drawing Gif' | Gif/rings.gif |   407.6 us |   170.7 us |   9.643 us |   1.00 |     0.00 | 16.6016 | 105.37 KB |
     'ImageSharp Gif' | Gif/rings.gif | 1,242.7 us | 2,101.4 us | 118.733 us |   3.05 |     0.25 |       - |   2.11 KB |
```


### DecodePNG

Baseline, PRE PR (Upstream / Master)  —

```
               Method |      TestImage |     Mean |     Error |    StdDev | Scaled | ScaledSD |   Gen 0 |   Gen 1 |   Gen 2 | Allocated |
--------------------- |--------------- |---------:|----------:|----------:|-------:|---------:|--------:|--------:|--------:|----------:|
 'System.Drawing Png' | Png/splash.png | 2.730 ms |  4.729 ms | 0.2672 ms |   1.00 |     0.00 | 74.2188 | 74.2188 | 74.2188 | 240.54 KB |
     'ImageSharp Png' | Png/splash.png | 7.718 ms | 13.086 ms | 0.7394 ms |   2.84 |     0.31 |       - |       - |       - |   97.2 KB |
```

**PR** —

using SpanHelper.Copy

```
               Method |      TestImage |     Mean |     Error |    StdDev | Scaled | ScaledSD |   Gen 0 |   Gen 1 |   Gen 2 | Allocated |
--------------------- |--------------- |---------:|----------:|----------:|-------:|---------:|--------:|--------:|--------:|----------:|
 'System.Drawing Png' | Png/splash.png | 2.695 ms |  3.219 ms | 0.1819 ms |   1.00 |     0.00 | 74.2188 | 74.2188 | 74.2188 | 240.54 KB |
     'ImageSharp Png' | Png/splash.png | 7.665 ms | 13.537 ms | 0.7649 ms |   2.85 |     0.28 |       - |       - |       - |   97.2 KB |
```

using Span.CopyTo
```
               Method |      TestImage |     Mean |    Error |    StdDev | Scaled | ScaledSD |   Gen 0 |   Gen 1 |   Gen 2 | Allocated |
--------------------- |--------------- |---------:|---------:|----------:|-------:|---------:|--------:|--------:|--------:|----------:|
 'System.Drawing Png' | Png/splash.png | 3.047 ms | 2.904 ms | 0.1641 ms |   1.00 |     0.00 | 74.2188 | 74.2188 | 74.2188 | 240.54 KB |
     'ImageSharp Png' | Png/splash.png | 6.939 ms | 1.496 ms | 0.0845 ms |   2.28 |     0.10 |       - |       - |       - |   97.2 KB |
```

Testing environment details:

| Intel Xeon CPU E5-2663 v3 2.80GHz, 1 CPU, 10 logical cores and 10 physical cores
| .NET Framework 4.7 (CLR 4.0.30319.42000), 64bit RyuJIT-v4.7.2633.0  **[slow span]**

File: src/ImageSharp/Image.LoadPixelData.cs (hunk 4 above)
Comment (iamcarbon) [reply src/ImageSharp/Image.LoadPixelData.cs]: The implicit cast doesn't seem to pick up in this case. Removing this comment (should revisit later).

**Pull request merged**